lib/std/fs.zig

```diff
@@ -1587,3 +1587,62 @@ test "" {
     _ = @import("fs/file.zig");
     _ = @import("fs/get_app_data_dir.zig");
 }
+
+// Custom read-by-line iterator with some FP-style secret sauce
+// Used for std.terminfo.Terminfo
+pub fn ReadLineIterator(comptime buffer_size: usize) type {
+    return struct {
+        const Self = @This();
+        const FifoType = std.fifo.LinearFifo(u8, std.fifo.LinearFifoBufferType{ .Static = buffer_size });
+        const FilterCallback = fn([]u8)bool;
+        const MapCallback = fn([]u8)[]const u8;
+
+        fifo: FifoType,
+        file: std.fs.File,
+        in_stream: std.fs.File.InStream,
+        filter_fn: FilterCallback,
+        map_fn: MapCallback,
+
+        pub fn init(file_path: []const u8) !Self {
+            const file = try std.fs.cwd().openFile(file_path, .{});
+            var in_stream = file.inStream();
+
+            return Self {
+                .fifo = FifoType.init(),
+                .file = file,
+                .in_stream = in_stream,
+                .filter_fn = struct { fn filter(line: []u8) bool { return true; } }.filter,
+                .map_fn = struct { fn map(line: []u8) []u8 { return line; } }.map,
+            };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.file.close();
+        }
+
+        pub fn setFilter(self: *Self, filter_fn: FilterCallback) void {
+            self.filter_fn = filter_fn;
+        }
+
+        pub fn setMap(self: *Self, map_fn: MapCallback) void {
+            self.map_fn = map_fn;
+        }
+
+        pub fn next(self: *Self) !?[]const u8 {
+            while (try (&self.in_stream.stream).readUntilDelimiterOrEof(self.fifo.buf[0..], '\n')) |line| {
```

Comment:
This is not how you use fifo: use `self.fifo.writableSlice(0)`  instead.
But really, you just want `BufferedInStream` (which may start using fifo internally in future)

Reply:
So... I decided to go with `BufferedInStream` since I want future-proof it, but I have a feeling I wasn't able to resolve this just yet, because I'm confused. :confused: